File: packages/agent-mesh/sdks/typescript/src/policy.ts
import { readFileSync } from 'fs';
import { PolicyRule } from './types';

export type PolicyDecision = 'allow' | 'deny' | 'review';

/**
 * Rule-based policy engine supporting action matching and conditional evaluation.
 */
export class PolicyEngine {
  private rules: PolicyRule[] = [];

  constructor(rules?: PolicyRule[]) {
    if (rules) {
      this.rules = [...rules];
    }
  }

  /** Load policy rules from a YAML file. */
  async loadFromYAML(yamlPath: string): Promise<void> {
    // Dynamic import so js-yaml remains an optional peer dep at runtime
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const yaml = require('js-yaml');
    const content = readFileSync(yamlPath, 'utf-8');
    const doc = yaml.load(content) as { rules?: PolicyRule[] };
    if (doc?.rules && Array.isArray(doc.rules)) {
      this.rules.push(...doc.rules);
    }
  }

  /**
   * Evaluate an action against loaded rules.
   *
   * Rules are evaluated in order; the first match wins.
   * If no rule matches the default decision is 'deny'.
   */
  evaluate(action: string, context: Record<string, unknown> = {}): PolicyDecision {
    for (const rule of this.rules) {
      if (this.matchAction(rule.action, action) && this.matchConditions(rule.conditions, context)) {
        return rule.effect;
      }
    }
    return 'deny'; // default-deny
  }

  /** Append a rule to the policy set. */
  addRule(rule: PolicyRule): void {
    this.rules.push(rule);
  }

  /** Return a snapshot of the current rules. */
  getRules(): readonly PolicyRule[] {
    return [...this.rules];
  }

  // ── Private helpers ──

  private matchAction(pattern: string, action: string): boolean {
    if (pattern === '*') return true;
    if (pattern.endsWith('.*')) {
      const prefix = pattern.slice(0, -2);
      return action === prefix || action.startsWith(prefix + '.');
    }
    return pattern === action;
  }

  private matchConditions(
    conditions: Record<string, unknown> | undefined,
    context: Record<string, unknown>,
  ): boolean {
    if (!conditions) return true;
    for (const [key, expected] of Object.entries(conditions)) {
      const actual = context[key];
      if (actual !== expected) return false;
    }
    return true;
  }
}

File: packages/agent-mesh/sdks/go/policy.go
package agentmesh

import (
	"os"
	"strings"
	"sync"

	"gopkg.in/yaml.v3"
)

// PolicyRule defines a single governance rule.
type PolicyRule struct {
	Action     string                 `json:"action" yaml:"action"`
	Effect     PolicyDecision         `json:"effect" yaml:"effect"`
	Conditions map[string]interface{} `json:"conditions,omitempty" yaml:"conditions,omitempty"`
}

// PolicyEngine evaluates actions against a set of rules.
type PolicyEngine struct {
	mu    sync.RWMutex
	rules []PolicyRule
}

// NewPolicyEngine creates a PolicyEngine with the supplied rules.
func NewPolicyEngine(rules []PolicyRule) *PolicyEngine {
	return &PolicyEngine{rules: rules}
}

// Evaluate returns the decision for the given action and context.
// Rules are evaluated in order; first match wins. Default is Deny.
func (pe *PolicyEngine) Evaluate(action string, context map[string]interface{}) PolicyDecision {
	pe.mu.RLock()
	defer pe.mu.RUnlock()

	for _, rule := range pe.rules {
		if matchAction(rule.Action, action) && matchConditions(rule.Conditions, context) {
			return rule.Effect
		}
	}
	return Deny
}

// LoadFromYAML loads rules from a YAML file, appending to existing rules.
func (pe *PolicyEngine) LoadFromYAML(path string) error {
	data, err := os.ReadFile(path)
	if err != nil {
		return err
	}

	var loaded struct {
		Rules []PolicyRule `yaml:"rules"`
	}
	if err := yaml.Unmarshal(data, &loaded); err != nil {
		return err
	}

	pe.mu.Lock()
	defer pe.mu.Unlock()
	pe.rules = append(pe.rules, loaded.Rules...)
	return nil
}

func matchAction(pattern, action string) bool {
	if pattern == "*" {
		return true
	}
	if strings.HasSuffix(pattern, ".*") {
		prefix := strings.TrimSuffix(pattern, ".*")
		return strings.HasPrefix(action, prefix+".")
	}
	return pattern == action
}

func matchConditions(conditions map[string]interface{}, context map[string]interface{}) bool {
	for key, expected := range conditions {
		actual, ok := context[key]
		if !ok {
			return false
		}
		if !valuesEqual(expected, actual) {
			return false
		}
	}
	return true
}

func valuesEqual(a, b interface{}) bool {
	switch av := a.(type) {
	case string:
		bv, ok := b.(string)
		return ok && av == bv
	case float64:
		bv, ok := b.(float64)
		return ok && av == bv
	case bool:
		bv, ok := b.(bool)
		return ok && av == bv
	default:
		return false
	}
}
